test/e2e/etcd_multi_node_test.go

```diff
@@ -0,0 +1,441 @@
+package e2e
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/gardener/etcd-druid/api/v1alpha1"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"github.com/gardener/gardener/pkg/utils/test/matchers"
+	"github.com/go-logr/logr"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	appsv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	v1 "k8s.io/api/coordination/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	k8s_labels "k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ = Describe("Etcd", func() {
+	Context("when multi-node is configured", func() {
+		var (
+			cl               client.Client
+			etcdName         string
+			storageContainer string
+			parentCtx        context.Context
+			provider         TestProvider
+			providers        []TestProvider
+			err              error
+		)
+
+		BeforeEach(func() {
+			parentCtx = context.Background()
+			providers, err = getProviders()
+			Expect(err).ToNot(HaveOccurred())
+			// take first provider
+			provider = providers[0]
+			cl, err = getKubernetesClient(kubeconfigPath)
+			Expect(err).ShouldNot(HaveOccurred())
+
+			etcdName = fmt.Sprintf("etcd-%s", provider.Name)
+
+			storageContainer = getEnvAndExpectNoError(envStorageContainer)
+
+			snapstoreProvider := provider.Storage.Provider
+			store, err := getSnapstore(string(snapstoreProvider), storageContainer, storePrefix)
+			Expect(err).ShouldNot(HaveOccurred())
+
+			// purge any existing backups in bucket
+			Expect(purgeSnapstore(store)).To(Succeed())
+			Expect(deployBackupSecret(parentCtx, cl, logger, provider, etcdNamespace, storageContainer))
+
+		})
+
+		AfterEach(func() {
+			// remove etcd objects if any old etcd objects exists.
+			purgeEtcd(parentCtx, cl, providers)
+		})
+
+		It("should perform etcd operations", func() {
+			ctx, cancelFunc := context.WithTimeout(parentCtx, 10*time.Minute)
+			defer cancelFunc()
+
+			etcd := getDefaultMultiNodeEtcd(etcdName, namespace, storageContainer, storePrefix, provider)
+			objLogger := logger.WithValues("etcd-multi-node", client.ObjectKeyFromObject(etcd))
+
+			By("Create etcd")
+			createAndCheckEtcd(ctx, cl, objLogger, etcd)
+
+			By("Hibernate etcd (Scale down from 3->0)")
+			hibernateAndCheckEtcd(ctx, cl, objLogger, etcd)
+
+			By("Wakeup etcd (Scale up from 0->3)")
+			wakeupEtcd(ctx, cl, objLogger, etcd)
+
+			By("Zero downtime rolling updates")
+			job := startEtcdZeroDownTimeValidatorJob(ctx, cl, etcd, "rolling-update")
+			// this defer ensures to remove the job, if test case breaks before deleting the job.
+			defer cleanUpTestHelperJob(ctx, cl, client.ObjectKeyFromObject(job))
+
+			ExpectWithOffset(1, cl.Get(ctx, client.ObjectKeyFromObject(etcd), etcd)).To(Succeed())
```

Comment:
```suggestion
			Expect(cl.Get(ctx, client.ObjectKeyFromObject(etcd), etcd)).To(Succeed())
```

`ExpectWithOffset` is only needed when called in a helper function.